### ISSUE TITLE
Completely broken with Chrome Beta

### ISSUE THREAD

**Issue by webxl** [dmarmor/epichrome#153]
Version 69.0.3497.23

When I run from the command line, I get errors like:

```
	file system sandbox blocked open() of '/Applications/Google Chrome.app/Contents/Versions/69.0.3497.23/Google Chrome Helper.app/Contents/MacOS/../../../Google Chrome Framework.framework/Google Chrome Framework'
dlopen /Applications/Google Chrome.app/Contents/Versions/69.0.3497.23/Google Chrome Helper.app/Contents/MacOS/../../../Google Chrome Framework.framework/Google Chrome Framework: dlopen(/Applications/Google Chrome.app/Contents/Versions/69.0.3497.23/Google Chrome Helper.app/Contents/MacOS/../../../Google Chrome Framework.framework/Google Chrome Framework, 261): no suitable image found.  Did find:
```
All tabs show "Can't open this page" or "Aw Snap!"

**Comment by ylluminate** [comment]
Same here all the way around.  Very disruptive situation - anyone find a solution so far?

**Comment by webxl** [comment]
@ylluminate I just went with my old solution, Chrome Canary. But not sure what to do if you're trying to use this for SSB.

**Comment by GrogHills** [comment]
Oh no. I'm seeing the issue too. Any tips?

Chrome Version 69.0.3497.23 (Official Build) beta (64-bit)

**Comment by jacopocolo** [comment]
For everyone with the same problem: I downgraded to Chrome 68 and everything is back to normal. It also looks like Chrome 69 is not available anymore? My Chrome shows version 68 as the most recent 🤷🏻‍♂️

**Comment by dmarmor** [comment]
Hi everyone! Thank you all for reporting this. I've gotten Epichrome working again with Chrome 69 and hopefully the new release will go up tonight. Please read the release notes, as I had to make some fairly major architectural changes to make it work, due to Chrome 69's heavy security.

**Comment by ylluminate** [comment]
Hey @dmarmor thank you so much for your effort in this area! I just attempted a 69 beta install and it gives a crash on each load (reload) running on 10.14 Beta (18A377a):
![image](https://cl.ly/66893cb804c7/Image%202018-09-03%20at%202.30.44%20AM.png)

Google Chrome Helper crash report: https://cl.ly/5705dd1a3877

**Comment by dmarmor** [comment]
Oh boy.

I’m still running 10.13 and it’s been working for me. Can you do me a favor and run one of your apps from the command-line, like so:

<your-app>/Contents/MacOS/Epichrome

And send me the stderr result?

**Comment by ylluminate** [comment]
So first upon execution, I got this:
Unable to update app to new version of Google Chrome. (Error filtering Info.plist file. (Failed to import the site module
Your PYTHONPATH points to a site-packages dir for Python 2.x but you are running Python 3.x!
     PYTHONPATH is currently: "/usr/local/lib/python2.7/site-packages:"
     You should `unset PYTHONPATH` to fix this.)) It may or may not run properly.

![dialogue of error](https://cl.ly/bd7a7dfcf1dd/Image%202018-09-03%20at%209.24.42%20AM.png)

Terminal output: https://cl.ly/3bb802a63475

Google Chrome Helper crash report again: https://cl.ly/c751b5587f78

**Comment by dmarmor** [comment]
Thanks for sending that. The good news is, it looks like problem is that it didn’t update properly, so there’s still hope 2.2.1 itself isn’t  broken under 10.14.

It looks like somehow Python 3 is installed at /usr/bin/python? Do you know if 10.14 uses python 3 as its default? Though if that was the case, I’m not sure why PYTHONPATH would still be pointing at the library for 2.7.

Is there any way you can get /usr/bin/python pointing at 2.7 again and see if that solves the problem?

**Comment by ylluminate** [comment]
Without looking at things closely before, I guess it could be due to MacPorts and Homebrew.  I've been moving everything over to MacPorts, but I still have `python` from both of those in the path; note the order:
```
/opt/local/bin/python # 3.6.6
/usr/local/bin/python # 2.7.15
/usr/bin/python # 2.7.10
```
So apparently Epichrome is honoring the path and using the first available instance of python vs the system's python. 

So `PYTHONPATH` is pointing at the Homebrew installation, oddly:
```
/usr/local/lib/python2.7/site-packages:
``` 

Hmm, so I need to think about how to handle this properly. I know a few other folks who are moving away from Homebrew to MacPorts now as well and they also use Epichrome so I need to figure out what to do here I guess without just going one way or the other fully. I suppose there must be a way to properly set PYTHONPATH for this situation that would resolve it.

**Comment by dmarmor** [comment]
That’s actually extra puzzling as I just fixed a relative path to python in one of the calls, so they all should be explicitly calling /usr/bin/python. Did you by chance download 2.2.0 in the few minutes it was up before I caught that and put out 2.2.1? If so, try the newer version and see if it fixes the problem. Fingers crossed. If not, I can’t figure out how it’s ending up running one of the other python instances.

**Comment by ylluminate** [comment]
Hmm, unfortunately I did get 2.2.1 and have the same problem: https://cl.ly/4a9a878fefb2

**Comment by dmarmor** [comment]
Well, something is definitely wrong, as that output is showing the Chrome engine exec still being called zzzzChromeEngine, which it no longer is in 2.2.1. I suspect something went wrong in your update process. Can you try creating an app from scratch with 2.2.1 to see if it works properly for you?

Then if you don't have a backup of your old app, it might be possible to roll back the version number in its config.sh and get it to try the update again. Or as a last resort, you could recreate the app and then swap in your existing user directory.

**Comment by ylluminate** [comment]
A new Epichrome app DOES work. Everything starts up just fine and Chrome 69 works in it. When I start an old app, I still get the dialogue: 
```
Unable to update app to new version of Google Chrome. (Error filtering Info.plist file. (Failed to import the site module
Your PYTHONPATH points to a site-packages dir for Python 2.x but you are running Python 3.x!
     PYTHONPATH is currently: "/usr/local/lib/python2.7/site-packages:"
     You should `unset PYTHONPATH` to fix this.)) It may or may not run properly.
```

So something is definitely wrong in the update process. When it finally does start up, after spewing all of the errors, it still does start with Chrome 68.x so it's giving a mish-mash of versions in some fashion.

**Comment by dmarmor** [comment]
Yeah, something about your environment is clearly stopping Epichrome from successfully updating your apps.

My recommendation would be to recreate those apps. If you have a lot of customization to your user profiles so that you don't want to start from scratch, there are a few strategies for preserving those, which are outlined in the Readme under "[Simple Method](https://github.com/dmarmor/epichrome/#simple-method)". If that won't work for you, it is possible to hand-update an app, but it's fairly annoying and error-prone. I can walk you through it if you need.

Hopefully with the relative-path-to-python bug fixed in 2.2.1 (and 2.2.2 which I've just released) this problem won't happen with future updates...

**Comment by ylluminate** [comment]
Yeah, there's no way I can recreate these apps. I've got at least 50 Epichrome apps. I'm going to have to figure out something else... 😢

**Comment by ylluminate** [comment]
Thanks... So yes, same error with 2.2.2: https://cl.ly/97bda2840a58

Could you just breakout a manual incantation for `/Applications/Epichrome.app/Contents/Resources/Runtime/Resources/Scripts/runtime.sh` that would reproduce the update in a manual fashion?  I'm assuming I need to know some args, etc.  Then I can try to hunt down what might be going on.  It almost seems like a permission issue, although I can't hunt it down yet.  `dtrace` isn't really shedding any light on it either yet so I've gotta get to where I can essentially execute each step in the shell manually...


Re: lots of apps... :)  I like me my Epichrome. I find it exceedingly useful to use for both projects and clients and other various webapps since I can keep everything cleanly consolidated.  I now also use an rsync script to copy instances between other workstations here at the office.  (The only problem with that is that Chrome is doing something funky that requires extensions + themes to be reinstalled... we've spent a little time troubleshooting that, but haven't been successful yet.)

**Comment by dmarmor** [comment]
This is so very weird--that log you sent in your last post *still* shows zzzzChromeEngine, which means it's definitely not 2.2.2. I just don't understand what's happening on your system.

Unfortunately you can't just run runtime.sh from the command-line without doing a lot of other work. It's really meant only to be run from within an app bundle.

I'm going to have to disappear from here for the next week or two. I hope you can get things working well enough to muddle through, maybe recreating a couple of your key apps? As soon as I can find some free time after this project I'm working on calms down, I can work on a remedial script to try and help you get your apps working again...

**Comment by ylluminate** [comment]
Hey thanks so much - no problem. I'm just going to give up on moving to 69 for now. I'll keep a backup of Chrome 68 for the time being - I use Safari for my main browser anyway and just use Epichrome apps for all my specifics so this should be able to hold without a lot of fuss.  I've got some pretty pressing things going on myself to keep me tied up - I was just hoping to resolve them before the next few days come upon me with some seriously taxing things.

The 69 update is nice in that it fixes some problems that Chrome has had floating around for a couple releases, but it's not earth shattering.

So yeah, I'll hang tight for when you can pop back onto this. Hopefully no one else has any trouble like me, especially if 10.14 is released here pretty soon!

**Comment by ylluminate** [comment]
Just to document this as well so I don't forget to look into it more later, I'm finding that when I start some apps up, even though I have 2.2.2 installed, I still get a download prompt for it in older Epichrome apps, eg:

![dialogue](https://cl.ly/248179982fac/Image%202018-09-04%20at%209.31.02%20AM.png)

And the `Info.plist` shows:
```
$ cat /Applications/Epichrome.app/Contents/Info.plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>CFBundleAllowMixedLocalizations</key>
	<true/>
	<key>CFBundleDevelopmentRegion</key>
	<string>English</string>
	<key>CFBundleExecutable</key>
	<string>applet</string>
	<key>CFBundleIconFile</key>
	<string>applet</string>
	<key>CFBundleInfoDictionaryVersion</key>
	<string>6.0</string>
	<key>CFBundleName</key>
	<string>Epichrome</string>
	<key>CFBundleIdentifier</key>
	<string>org.epichrome.Epichrome</string>
	<key>CFBundlePackageType</key>
	<string>APPL</string>
	<key>CFBundleShortVersionString</key>
	<string>2.2.2</string>
	<key>CFBundleSignature</key>
	<string>aplt</string>
	<key>LSMinimumSystemVersionByArchitecture</key>
	<dict>
		<key>x86_64</key>
		<string>10.6</string>
	</dict>
	<key>LSRequiresCarbon</key>
	<true/>
</dict>
</plist>
```

**Comment by dmarmor** [comment]
Is that one of your apps that you've tried to update but wasn't working under Chrome 69? Can you check the config.sh and see if the version number is 2.2.2 in there? If not, this is probably just more fallout from the incomplete update, and it's looking at that version number and seeing that it's not the latest.

If it *is* 2.2.2 in config.sh, then it looks like I've got a bug in the version-checking code and will have to work on that when I'm back in action.

Thanks for all the diagnostic help on this! I hope we can get you back in action relatively soon...

**Comment by ylluminate** [comment]
So on that particular app noted above (it's an app dedicated to Racket programming so I keep all of those topics consolidated and the browser opens back up where I left off), the `config.sh` shows:
```
# config.sh -- autogenerated Tue Sep  4 09:33:00 EDT 2018

CFBundleDisplayName=Racket
CFBundleName=Racket
CFBundleIdentifier=org.epichrome.app.Racket701
SSBVersion=2.1.20
SSBUpdateCheckDate=1536672248
SSBUpdateCheckVersion=2.1.20
SSBProfilePath=Library/Application\ Support/Epichrome/Apps/Racket701
SSBChromePath=/Applications/Google\ Chrome.app
SSBChromeVersion=68.0.3440.106
SSBRegisterBrowser=Yes
SSBCustomIcon=Yes
SSBFirstRunSinceVersion=''
SSBHostInstallError=''
SSBCommandLine=( )
```

And I can't stress enough: thank you for being such a butt kicker on this project.  Epichrome is a huge gift and it's unfortunate that its not more widely known in terms of its power and the SANITY it can bring to one's life.  I sure wish this was possible with Firefox as I think it could be even better, but nonetheless this is a savior when you take advantage of its power - especially for folks like me who just have such varied lives, responsibilities and interests.

**Comment by henderea** [comment]
I was having issues getting it to offer updating the runtime after installing the new version of epichrome.  I ended up discovering that the issue was an old copy of Epichrome.app elsewhere on my filesystem, which one of the mdfind commands was picking up on.

So here's the solution that worked for me:

If either of these commands show anything, delete the files they point to:

```shell
mdfind "kMDItemCFBundleIdentifier == 'org.epichrome.builder'"
```
```shell
mdfind "kMDItemCFBundleIdentifier == 'com.dmarmor.MakeChromeSSB'"
```

Make sure to check both.  After doing this, try relaunching the epichrome app and it will hopefully offer to update the runtime.

**Comment by GrogHills** [comment]
Thank you so much for taking the time to update epichrome with the latest chrome version. I just updated and everything is working great! I love epichrome. Thank you!

**Comment by ylluminate** [comment]
@henderea good call and good idea!  
```
$ mdfind "kMDItemCFBundleIdentifier == 'com.dmarmor.MakeChromeSSB'"
```
revealed 3 instances of `Epichrome.app` from both my ~/Applications/ and 2 from other backup volumes.  After removing these and restarting, I initially got a quick (only was up for about 1-1.5 seconds) "Profile error occurred" bar, but then after I put in Chrome 69, boom, it updated and it now works!

So @dmarmor fantastic news in this area with it working now.

**Comment by henderea** [comment]
@ylluminate to be clear, the issue is any Epichrome.app indexed by Spotlight (other than the main copy you updated, obviously).  So if you don't have Spotlight enabled on your backup volume, copies of the app on there won't matter.

I don't know what the `md` in `mdfind` is for, but that command uses the Spotlight index.  Somehow, all of the processes and shell commands for Spotlight use `md` in their name.  Maybe that stands for "metadata".  Anyway, the commands just look in the Spotlight index, so turning it off for a backup volume will prevent that volume from being an issue.

**Comment by dmarmor** [comment]
Thank you @henderea for noticing the mdfind problem. I've stripped out some legacy code and changed the way mdfind is used so it should now always find only the latest installed version.

@ylluminate -- I'm really glad getting rid of the old versions has helped! Has that solved all your updating problems, or are you still having trouble with some of the other updating issues?

**Comment by ylluminate** [comment]
@dmarmor it seems all is good now as far as updating goes!

**Comment by dmarmor** [comment]
That's great news! Give 2.2.3 a try and if all is well, fingers crossed we can close this issue...

**Comment by ylluminate** [comment]
So @dmarmor another workstation was updated here and the user hit a wall.  There was an Epichrome.app instance on another volume again and he struggled with it for about an hour.  After I heard about it I executed `mdfind "kMDItemCFBundleIdentifier == 'com.dmarmor.MakeChromeSSB'"` and found that other instance and upon removing it (`sudo rm -fr`'ing it) the Epichrome apps started up and updated as per normal after that... Seems that other instances are still being problems for some folks.
